From d2f1915b6e5b536d7174ce6ca4a7d6e4a045f00d Mon Sep 17 00:00:00 2001
From: Yanjiang Jin <yanjiang.jin@windriver.com>
Date: Fri, 11 May 2012 11:06:16 +0800
Subject: [PATCH 32/47] nlm_xlp_64_be: fix the error link status of GE port

Correct the wrong PHY AD which is stored in a global
struct. Then we can get the actual link status.

Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 arch/mips/include/asm/netlogic/hal/nlm_nae.h |    2 ++
 drivers/net/nae/xlp_hw.c                     |    5 ++++-
 drivers/net/nae/xlp_nae.c                    |    2 +-
 3 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/arch/mips/include/asm/netlogic/hal/nlm_nae.h b/arch/mips/include/asm/netlogic/hal/nlm_nae.h
index ec8e9a8..4b6c4ee 100644
--- a/arch/mips/include/asm/netlogic/hal/nlm_nae.h
+++ b/arch/mips/include/asm/netlogic/hal/nlm_nae.h
@@ -41,6 +41,8 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #define NAE_RESET_DONE		0x20
 #define NAE_INIT_VALID		0x40
 
+#define INVALID_PHY_STATUS	0xffff
+
 struct nlm_hal_nae_port {
         int  valid;
         int  mgmt;
diff --git a/drivers/net/nae/xlp_hw.c b/drivers/net/nae/xlp_hw.c
index 3fbd9d7..6f57c99 100644
--- a/drivers/net/nae/xlp_hw.c
+++ b/drivers/net/nae/xlp_hw.c
@@ -264,7 +264,10 @@ static u32 xlp_get_link(struct net_device *dev)
 
 	spin_unlock_irqrestore(&priv->lock, flags);
 
-	if(mii_status & BMSR_LSTATUS)
+	/* If the NAE device which is defined by user in the DTS
+	doesn't fit the actual situation, we always get PHY
+	status as 0xffff*/
+	if(mii_status != INVALID_PHY_STATUS && mii_status & BMSR_LSTATUS)
 		return 1;
 	return 0;
 }
diff --git a/drivers/net/nae/xlp_nae.c b/drivers/net/nae/xlp_nae.c
index 9a3e6dd..cfed0d7 100644
--- a/drivers/net/nae/xlp_nae.c
+++ b/drivers/net/nae/xlp_nae.c
@@ -959,7 +959,7 @@ static void nlm_xlp_nae_init(void)
 			switch(nae_cfg->ports[i].iftype) {
 				case SGMII_IF:
 					priv->index = nae_cfg->ports[i].hw_port_id & 0x3;
-					priv->phy.addr = nae_cfg->ports[i].hw_port_id;
+					priv->phy.addr = nae_cfg->ports[i].ext_phy_addr;
 					break;
 				case XAUI_IF:
 					priv->index = XGMAC;
-- 
1.7.0

